src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java

```diff
@@ -1,99 +1,256 @@
 package org.broadinstitute.hellbender.tools.spark.sv;
 
 import com.google.common.annotations.VisibleForTesting;
-import htsjdk.samtools.util.Histogram;
-import htsjdk.samtools.SAMReadGroupRecord;
 import htsjdk.samtools.SamPairUtil;
+import htsjdk.samtools.SAMFileHeader;
 import htsjdk.samtools.metrics.MetricsFile;
-import org.apache.spark.api.java.JavaRDD;
-import org.apache.commons.math3.stat.descriptive.DescriptiveStatistics;
-import org.broadinstitute.hellbender.tools.picard.analysis.InsertSizeMetrics;
+import htsjdk.samtools.util.Histogram;
 import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.metrics.MetricAccumulationLevel;
+import org.broadinstitute.hellbender.tools.picard.analysis.InsertSizeMetrics;
 
+import org.apache.spark.api.java.JavaRDD;
+
+import scala.Tuple2;
+
+import java.util.*;
+import java.util.function.Function;
+import java.util.stream.StreamSupport;
+import java.util.stream.Collectors;
 import java.io.Serializable;
-import java.util.List;
-import java.util.Map;
-import java.util.TreeMap;
-import java.util.SortedMap;
 
 /**
  * Worker class to collect insert size metrics.
+ * TODO: filter reads based on length value (if too large), and/or minimum_pct like in Picard.
+ * TODO: is it desired that mapping quality is collected as well?
+ * TODO: handle higher-case lower-case group name when asking for a particular group's stats
  */
-public final class InsertSizeMetricsCollectorSpark implements Serializable{
+public final class InsertSizeMetricsCollectorSpark implements Serializable {
     private static final long serialVersionUID = 1L;
 
-    private final InsertSizeMetrics metrics = new InsertSizeMetrics();
-    private Histogram<Integer> htslibHistogram;
+    // histograms and metrics are stored in an order that coarser level groups appear in front of finer level
+    //     (library is a level coarser than read group level)
+    //     within the same level, entries are sorted by the GroupInfo alphanumerically.
+    // the accompanying map helps retrieving the information
+    // when a particular group's name in String format is given.
+    private List<SortedMap<GroupInfo, Map<SamPairUtil.PairOrientation, Tuple2<Histogram<Integer>, InsertSizeMetrics>>>> histogramsAndMetrics = new LinkedList<>();
+    private Map<String, Integer> mapFromGroupNameToIndexInList = new HashMap<>();
```

Comment:
you could call this map "listIndexByGroupName" -- a little shorter but I think conveys the same info. Your call.